### ISSUE TITLE
Is it possible to implement sparse mean squared error loss in DeepChem?

### ISSUE THREAD

**Issue by zhouhao-learning** [deepchem/deepchem#1356]
I am using deepchem's GraphConv model to train a drug target data. This is a regression problem, but there are many missing values in my data. I don't want to discard them, so I want to customize a MSE loss function, but I can't He successfully applied to the model
This is my code:

```
# coding=utf-8
from __future__ import absolute_import
from __future__ import print_function
import sys
import os
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from deepchem.models.tensorgraph.models.graph_models import GraphConvTensorGraph
import tensorflow as tf
from deepchem.models.tensorgraph.layers import Layer


class Mask_value_loss(Layer):
	def __init__(self, in_layers=None, mask_value=None,**kwargs):
		self.mask_value = mask_value
		super(Mask_value_loss, self).__init__(in_layers, **kwargs)
		try:
			shape1 = self.in_layers[0].shape
			shape2 = self.in_layers[1].shape
			if shape1[0] is None:
				self._shape = (shape2[0],)
			else:
				self._shape = (shape1[0],)
		except:
			pass

	def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
		inputs = self._get_input_tensors(in_layers, True)
		guess, label = inputs[0], inputs[1]
		mask_true = tf.cast(tf.not_equal(label,self.mask_value), tf.float32)
		masked_squared_error = tf.square(mask_true * (label - guess))
		if len(inputs) > 2:
			masked_squared_error *= inputs[2]
		out_tensor = tf.reduce_sum(masked_squared_error, axis=list(range(1, len(label._shape))))/ \
		             tf.reduce_sum(mask_true, axis=list(range(1, len(label._shape))))

		if set_tensors:
			self.out_tensor = out_tensor
		return out_tensor


def run():
	model = GraphConvTensorGraph(n_tasks=2,
						graph_conv_layers =[64],
						dense_layer_size=128,
						dropout=0.2,
						batch_size=50,
						learning_rate=0.001,
						mode="regression",
						model_dir=os.path.dirname(__file__))
	losses = Mask_value_loss()
	model.set_loss(losses)
	model.build()
	model.save()
	return model

if __name__ == "__main__":
	run()
```
The purpose of my custom loss function is to mask out the missing values in the data. My idea comes from keras. The following part of the code is my custom MSE using tensorflow, which is calculated by masking a specified value:

```
import tensorflow as tf


def masked_mse(mask_value):
    def f(y_true, y_pred):      
        mask_true = tf.cast(tf.not_equal(y_true, mask_value), tf.float32)
        masked_squared_error = tf.square(mask_true * (y_true - y_pred))
        masked_mse = tf.reduce_sum(masked_squared_error, axis=-1) / tf.reduce_sum(mask_true, axis=-1)
        return masked_mse
    return f
```
Assume that the input is like this：
```
y_pred = tf.constant([[1, 1, 1, 1],
                     [1, 1, 1, 3],
                     [1, 1, 1, 3],
                     [1, 1, 1, 3],
                     [1, 1, 1, 3],
                     [1, 1, 1, 3]], dtype=tf.float32)

y_true = tf.constant([[1, 1, 1, 1],
                     [1, 1, 1, 1],
                     [-1, 1, 1, 1],
                     [-1, -1, 1, 1],
                     [-1, -1, -1, 1],
                     [-1, -1, -1, -1]], dtype=tf.float32)

sess = tf.Session()
with sess.as_default():
    true = sess.run(y_true)
    pred = sess.run(y_pred)

    loss = sess.run(masked_mse(-1)(y_true, y_pred))

for i in range(true.shape[0]):
    print(true[i], pred[i], loss[i], sep='\t')
```
We can get such an output:

```
[1. 1. 1. 1.]	[1. 1. 1. 1.]	0.0
[1. 1. 1. 1.]	[1. 1. 1. 3.]	1.0
[-1.  1.  1.  1.]	[1. 1. 1. 3.]	1.3333334
[-1. -1.  1.  1.]	[1. 1. 1. 3.]	2.0
[-1. -1. -1.  1.]	[1. 1. 1. 3.]	4.0
[-1. -1. -1. -1.]	[1. 1. 1. 3.]	nan
```
But I can't customize such a loss function in DeepChem. Can you give me some pointers?

**Comment by peastman** [comment]
Masking out missing values is a standard feature in DeepChem.  All the loss functions support it automatically.

**Comment by zhouhao-learning** [comment]
@lilleswing 
What you said is that I need to set the missing value weight to 0? Is there a specific example for reference? In addition, your colleague @peastman  said that the loss loss function in DeepChem has the function of automatically masking missing values. Is this uniform?

**Comment by zhouhao-learning** [comment]
@peastman 
In other words, I don't have to write a loss function to mask missing values. The loss I got after training is already the loss value after removing the missing value, so I don't have to think about it again.

**Comment by peastman** [comment]
More specifically, the loss for every sample/task pair is multiplied by the corresponding entry in the dataset's weight matrix.  If you set an element of that matrix to 0, the corresponding sample/task will not contribute to the loss.  This is the standard way of marking missing data.  The weight matrix can also be used to correct for unbalanced data.

**Comment by Running-z** [comment]
@peastman 
Are you talking about the BalancingTransformer function? This should apply to the classification task, I am facing regression data.Does the regression data have a corresponding method for dealing with missing values in DeepChem?

**Comment by peastman** [comment]
Every dataset has a weight matrix.  Every loss function, whether it's for regression or classification, multiplies the losses by that weight matrix.  When you create your dataset, mark missing values by setting their weights to 0.
